### ISSUE TITLE
"PKIX path building failed" error, even after making sure AWS certs exist in cacerts files

### ISSUE THREAD

**Issue by tsibiski** [aws/aws-sdk-java#2145]
Hi,

I am simply trying to connect to the Amazon S3 bucket from my Java app. I have been using a C# framework with zero issues, using identical buckets and credentials for months now. It was requested that this framework be converted to Java.

Everything else has been fine, but Amazon S3 refuses to work. I always get 

`PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target`

I have googled the heck out of this issue, and reviewed dozens of previous cases where people get this issue. I downloaded the cert from Amazon's AWS Java cert page. I used Keytool to combine it with the jdk & jre /lib/security/cacerts files. In both cases, it said that the cert already existed in the files.

Next, I downloaded a cert after navigating directly to the amazon bucket I am connecting to. That cert did not seem to exist, and I added it to the cacert file using keytool commands. I then read the certs to a file, and confirmed that both certs existed in the file, and everything matched what I was seeing online.

Neither of these approaches worked. I copied the cert to EVERY directory that I could find with a Java cacerts file on my C drive. This includes the IntelliJ directory. I tested my code after each one was replaced. When it didn't work, I reverted each one. Still, I get the cert error. Then I simply replaced them all at once with the new cacerts file. Still the same error.

I tried over a dozen different versions of the Amazon AWS sdk's in my pom.xml file. I tried 3 different versions of Java, 3 different JDKs, and 3 different JRE's. 

Next I tested on ANOTHER computer and replicated everything above. I still get the same errors. What is happening? I've spent two weeks on this. I am out of ideas. No one seems to know what is wrong on my side, and all suggestions are the same as what I found by googling the error.

I've debugged into the SDK and found that the error is being thrown in this file:

C:\Users\XXXXXX\.m2\repository\com\amazonaws\aws-java-sdk-core\1.11.670\aws-java-sdk-core-1.11.670.jar!\com\amazonaws\http\conn\ **ClientConnectionManagerFactory.class**

**Line 48** _(Version 1.11.670 of the sdk)_

I further tracked this down to the apache file 
C:\Users\XXXXXX\.m2\repository\org\apache\httpcomponents\httpclient\4.5.5\httpclient-4.5.5.jar!\org\apache\http\conn\ssl\ **SSLConnectionSocketFactory.class**

**Line 204**  _(Version 4.5.5 of the jar)_

Debugging terminates here and the error bubbles up the stack from there. I am unable to step into or over the code any further that that point.

**Comment by zoewangg** [comment]
This is most likely to due to missing Amazon Root CA's in your trust store.

You can check what certificates are installed in your trust store with the following command:

```
keytool -list -v /path/to/keystore
```

And ensure the certificates from this page are present: https://aws.amazon.com/blogs/security/how-to-prepare-for-aws-move-to-its-own-certificate-authority/

    CN=Amazon Root CA 1,O=Amazon,C=US
    CN=Amazon Root CA 2,O=Amazon,C=US
    CN=Amazon Root CA 3,O=Amazon,C=US
    CN=Amazon Root CA 4,O=Amazon,C=US
    CN=Starfield Services Root Certificate Authority – G2,O=Starfield Technologies, Inc.,L=Scottsdale,ST=Arizona,C=US
    Starfield Class 2 Certification Authority

**Comment by tsibiski** [comment]
Thanks, but I already confirmed all certs were present.

The issue was due to an application called Netskope. Netskope was intercepting the AWS certs from Amazon, modifying them, calling themselves the source of the cert, and it was not matching the cacerts file certs, breaking everything.

If you come across this issue yourself, look for Netskope, or a similar security app, and disable it. Then test again. In IntelliJ, I added "-Djavax.net.debug=all" to my Run Configurations and "VM options" field to get more detailed info. It was in that debug output that I saw Netskope was fundamentally changing the certs and breaking them.
